File: examples/python/run_open_loop_demo.py
"""Open-loop simulation demo for the reduced stay-cable model."""

from pathlib import Path
import sys

import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "src" / "python"))

from cable_model import CableParameters, cable_state_rhs


def main() -> None:
    """Run a reproducible open-loop simulation."""

    params = CableParameters()
    t_span = (0.0, 120.0)
    t_eval = [0.05 * i for i in range(int(t_span[1] / 0.05) + 1)]
    initial_state = [0.02, 0.0]

    solution = solve_ivp(
        fun=lambda t, x: cable_state_rhs(t, x, params),
        t_span=t_span,
        y0=initial_state,
        t_eval=t_eval,
        rtol=1e-8,
        atol=1e-10,
    )

    if not solution.success:
        raise RuntimeError(solution.message)

    plt.figure()
    plt.plot(solution.t, solution.y[0])
    plt.xlabel("Time (s)")
    plt.ylabel("Modal displacement")
    plt.title("Open-loop stay-cable response")
    plt.grid(True)
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: examples/matlab/run_open_loop_demo.m
% RUN_OPEN_LOOP_DEMO  Open-loop simulation demo for a reduced stay-cable model.
%
% This script is intentionally compact. It provides a reproducible baseline
% for later controller comparisons such as LQR, MPC, and output-feedback MPC.

clear; clc; close all;

repoRoot = fileparts(fileparts(fileparts(mfilename('fullpath'))));
addpath(genpath(fullfile(repoRoot, 'src', 'matlab')));

p.omega = 1.6425;          % rad/s, first modal circular frequency
p.xi = 0.001;              % damping ratio
p.beta2 = 0.0;             % quadratic nonlinear coefficient
p.beta3 = 0.0;             % cubic nonlinear coefficient
p.excitation_amp = 0.02;   % normalized excitation amplitude
p.excitation_freq = 1.6425;% rad/s

x0 = [0.02; 0.0];          % initial modal displacement and velocity
tspan = [0, 120];

odeOpt = odeset('RelTol', 1e-8, 'AbsTol', 1e-10);
[t, x] = ode45(@(t, x) cable_rhs(t, x, p, 0.0), tspan, x0, odeOpt);

figure('Color', 'w');
plot(t, x(:, 1), 'LineWidth', 1.2);
grid on;
xlabel('Time (s)');
ylabel('Modal displacement');
title('Open-loop stay-cable response');
